simplemma/language_detector.py

```diff
@@ -10,24 +10,44 @@
 from .dictionary_factory import DictionaryFactory
 from .tokenizer import Tokenizer
 
+
 SPLIT_INPUT = re.compile(r"[^\W\d_]{3,}")
+RELAXED_SPLIT_INPUT = re.compile(r"[\w-]{3,}")
 
 
 class TokenSampler:
-    __slots__ = ["tokenizer"]
+    __slots__ = ["capitalized_threshold", "max_tokens", "tokenizer"]
 
-    def __init__(self, tokenizer: Tokenizer = Tokenizer(SPLIT_INPUT)) -> None:
+    def __init__(
+        self,
+        tokenizer: Tokenizer = Tokenizer(SPLIT_INPUT),
+        max_tokens: int = 1000,
+        capitalized_threshold: float = 0.8,
+    ) -> None:
         self.tokenizer = tokenizer
+        self.max_tokens = max_tokens
+        self.capitalized_threshold = capitalized_threshold
 
     def sample_tokens(self, text: str) -> List[str]:
         """Extract potential words, scramble them, extract the most frequent,
         some of the rest, and return at most 1000 tokens."""
-        # generator expression to split the text
-        counter = Counter(
-            token for token in self.tokenizer.split_text(text) if not token[0].isupper()
-        )
 
-        return [item[0] for item in counter.most_common(1000)]
+        counter = Counter(token for token in self.tokenizer.split_text(text))
+
+        deletions = [token for token in counter if token[0].isupper()]
+        if len(deletions) < self.capitalized_threshold * len(counter):
```

Comment:
This can delete up to 80% of the text.
Is that what you wanted? Seems like a lot.

Btw, you could also  sample here the max between deletions & capitalized_threshold and delete that, so you can delete something even if not everything.

Reply:
Sampling between both operations can be detrimental in case more words are needed in the end, and it may also be slower. I'm deciding against it for now.